### PR TITLE
engine-schema: fix finding guestos mapping with parent version

### LINKED ISSUE #8412: VMware: VM ingestion broken in case of dynamically handled hypervisor/guest OS versions.
ACS 4.19 RC1

VMware ingestion is broken when testing with VMware 8.0.0.2as the hypervisor and Rocky Linux (8 or 9) - this is a consequence of changes implemented in https://github.com/apache/cloudstack/pull/7380:

- if no record for 8.0.0.2 in the hypervisor_capabilities table, then use "capabilities" of the "parent" version which is VMware 8.0; 
- if no records in the guest_os_hypervisor table, then use the records/mappings of the "parent" version - from 8.0
- this all works fine during VM deployment
- this does not work during VMware ingestion, where we do reverse-mapping, and code seems to look for a DB record for "rockylinux_64Guest" for version 8.0.0.2 in the guest_os_hypervisor table - and this record does not exist

The logic of "mapping minor version guest OS mappings, if missing, to a parent version mappings" - sounds fine, as long as:

- It's supported by all ACS features/code where we try to use guest OS hypervisor mappings (this now includes VMware VM ingestion that needs to be fixed, but there might be other places)
- there is detailed logging similar to "The guest OS mappings for XXX was not found for the hypervisor version YYYYY (8.0.0.2), looking for parent version mappings (8.0), found, using that mapping etc" - something similar to this, needs to be very verbose, otherwise troubleshooting will be impossible.

For the hypervisor capabilities, I would propose the following:

- Revert the logic of dynamically "assuming" that the minor version has the same capabilities as the major version (in fugure, VMware e.g. 8.5 version will certainly NOT have the same capabilities as the 8.0 version, it would have more capabilities) - so revert the code and recreated the needed DB records in the hypervisor_capabilities table
- Worst case, even though it's a bad solution - leave it as it is, but add a loooot of logging, similar to the guest OS mappings logging (not found this, using that version capabilities)

### PR DESCRIPTION
**Pull request by shwstppr**
### Description

Fixes #8412

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by shwstppr** [comment]
@blueorangutan package

File: engine/schema/src/main/java/com/cloud/storage/dao/GuestOSHypervisorDaoImpl.java [hunk 1]
```diff
@@ -100,6 +103,21 @@ public GuestOSHypervisorVO findByOsIdAndHypervisor(long guestOsId, String hyperv
...
 
+    @Override
+    public GuestOSHypervisorVO findByOsIdAndHypervisor(long guestOsId, String hypervisorType, String hypervisorVersion) {
+        GuestOSHypervisorVO mapping = findByOsIdAndHypervisorInternal(guestOsId, hypervisorType, hypervisorVersion);
+        if (mapping != null || !Hypervisor.HypervisorType.VMware.toString().equals(hypervisorType)) {
+            return mapping;
+        }
+        String parentVersion = CloudStackVersion.getVMwareParentVersion(hypervisorVersion);
+        if (parentVersion == null) {
+            return null;
```
Comment (harikrishna-patnala) [inline engine/schema/src/main/java/com/cloud/storage/dao/GuestOSHypervisorDaoImpl.java:114]: Can we have a log message here stating that the parent is not found, so returning null.

File: engine/schema/src/main/java/com/cloud/storage/dao/GuestOSHypervisorDaoImpl.java [hunk 2]
```diff
@@ -138,6 +155,21 @@ public GuestOSHypervisorVO findByOsNameAndHypervisor(String guestOsName, String
...
 
+    @Override
+    public GuestOSHypervisorVO findByOsNameAndHypervisor(String guestOsName, String hypervisorType, String hypervisorVersion) {
+        GuestOSHypervisorVO mapping = findByOsNameAndHypervisorInternal(guestOsName, hypervisorType, hypervisorVersion);
+        if (mapping != null || !Hypervisor.HypervisorType.VMware.toString().equals(hypervisorType)) {
+            return mapping;
+        }
+        String parentVersion = CloudStackVersion.getVMwareParentVersion(hypervisorVersion);
+        if (parentVersion == null) {
+            return null;
```
Comment (harikrishna-patnala) [inline engine/schema/src/main/java/com/cloud/storage/dao/GuestOSHypervisorDaoImpl.java:166]: same here please, some log message

**Comment by shwstppr** [comment]
@blueorangutan package

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Comment by shwstppr** [comment]
@blueorangutan test rocky8 vmware-67u3

**Comment by shwstppr** [comment]
@blueorangutan test

**Review by harikrishna-patnala — APPROVED** [review]
code LGTM. I see this definetely fix the VMware VM ingestion part. We can think about support of parent version for hypervisor capabilities separately.

**Review by weizhouapache — APPROVED** [review]
code lgtm

**Comment by weizhouapache** [comment]
@shwstppr 
Should it be re-targeted to 4.18?

**Comment by shwstppr** [comment]
> @shwstppr Should it be re-targeted to 4.18?

will make the change

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test matrix

**Comment by shwstppr** [comment]
Tested `Rocky Linux 9` VM import with VMware 8.0.0.2 hypervisor.

```
(localcloud) 🐱 > list hypervisorcapabilities hypervisor=VMware filter=id,hypervisor,hypervisorversion,
{
  "count": 19,
  "hypervisorCapabilities": [
    {
      "hypervisor": "VMware",
      "hypervisorversion": "default",
      "id": "8"
    },
    {
      "hypervisor": "VMware",
      "hypervisorversion": "4.0",
      "id": "9"
    },
    {
      "hypervisor": "VMware",
      "hypervisorversion": "4.1",
      "id": "10"
    },
    {
      "hypervisor": "VMware",
      "hypervisorversion": "5.0",
      "id": "11"
    },
    {
      "hypervisor": "VMware",
      "hypervisorversion": "5.1",
      "id": "12"
    },
    {
      "hypervisor": "VMware",
      "hypervisorversion": "5.5",
      "id": "c57558bf-abc1-11ee-8957-1e00da000152"
    },
    {
      "hypervisor": "VMware",
      "hypervisorversion": "6.0",
      "id": "cbaae7e7-abc1-11ee-8957-1e00da000152"
    },
    {
      "hypervisor": "VMware",
      "hypervisorversion": "6.5",
      "id": "cd741f46-abc1-11ee-8957-1e00da000152"
    },
    {
      "hypervisor": "VMware",
      "hypervisorversion": "6.7",
      "id": "cf92dd65-abc1-11ee-8957-1e00da000152"
    },
    {
      "hypervisor": "VMware",
      "hypervisorversion": "6.7.1",
      "id": "cf92e9d9-abc1-11ee-8957-1e00da000152"
    },
    {
      "hypervisor": "VMware",
      "hypervisorversion": "6.7.2",
      "id": "cf92f45c-abc1-11ee-8957-1e00da000152"
    },
    {
      "hypervisor": "VMware",
      "hypervisorversion": "6.7.3",
      "id": "cf92fed5-abc1-11ee-8957-1e00da000152"
    },
    {
      "hypervisor": "VMware",
      "hypervisorversion": "7.0",
      "id": "d14750d0-abc1-11ee-8957-1e00da000152"
    },
    {
      "hypervisor": "VMware",
      "hypervisorversion": "7.0.1.0",
      "id": "d1487eae-abc1-11ee-8957-1e00da000152"
    },
    {
      "hypervisor": "VMware",
      "hypervisorversion": "7.0.2.0",
      "id": "d2c839eb-abc1-11ee-8957-1e00da000152"
    },
    {
      "hypervisor": "VMware",
      "hypervisorversion": "7.0.3.0",
      "id": "d2c96de9-abc1-11ee-8957-1e00da000152"
    },
    {
      "hypervisor": "VMware",
      "hypervisorversion": "8.0",
      "id": "d414b048-abc1-11ee-8957-1e00da000152"
    },
    {
      "hypervisor": "VMware",
      "hypervisorversion": "8.0.0.1",
      "id": "d414bc3b-abc1-11ee-8957-1e00da000152"
    },
    {
      "hypervisor": "VMware",
      "hypervisorversion": "8.0.1",
      "id": "d67c19b1-abc1-11ee-8957-1e00da000152"
    }
  ]
}
```

Import in UI,


https://github.com/apache/cloudstack/assets/153340/d34ca524-acbb-42d6-a4a2-972d99afe270


In Logs,
```
2024-01-08 09:23:51,845 DEBUG [c.c.a.t.Request] (API-Job-Executor-31:ctx-dfff55c7 job-38 ctx-4a1de5f8) (logid:7f598e47) Seq 1-5056135006653522013: Received:  { Ans: , MgmtId: 167780543, via: 1(10.0.35.115), Ver: v1, Flags: 10, { GetUnmanagedInstancesAnswer } }
2024-01-08 09:23:51,845 DEBUG [c.c.a.m.AgentManagerImpl] (API-Job-Executor-31:ctx-dfff55c7 job-38 ctx-4a1de5f8) (logid:7f598e47) Details from executing class com.cloud.agent.api.GetUnmanagedInstancesCommand: 
2024-01-08 09:23:51,854 DEBUG [c.c.s.d.GuestOSHypervisorDaoImpl] (API-Job-Executor-31:ctx-dfff55c7 job-38 ctx-4a1de5f8) (logid:7f598e47) Mapping for guest OS ID: rockylinux_64Guest for hypervisor: VMware with version: 8.0.0.2 can not be found. Trying to find one for the parent version: 8.0
2024-01-08 09:23:51,862 DEBUG [c.c.u.AccountManagerImpl] (API-Job-Executor-31:ctx-dfff55c7 job-38 ctx-4a1de5f8) (logid:7f598e47) Access granted to Account [{"accountName":"admin","id":2,"uuid":"e06f184a-abc1-11ee-8957-1e00da000152"}] to Service offering {"id":1,"name":"Small Instance","uuid":"5aebc1b7-5e3a-405f-98cf-dec841af8453"}. by AffinityGroupAccessChecker

```

**Comment by weizhouapache** [comment]
> Tested `Rocky Linux 9` VM import with VMware 8.0.0.2 hypervisor.
> 
> ```
> (localcloud) 🐱 > list hypervisorcapabilities hypervisor=VMware filter=id,hypervisor,hypervisorversion,
> {
>   "count": 19,
>   "hypervisorCapabilities": [
>     {
>       "hypervisor": "VMware",
>       "hypervisorversion": "default",
>       "id": "8"
>     },
>     {
>       "hypervisor": "VMware",
>       "hypervisorversion": "4.0",
>       "id": "9"
>     },
>     {
>       "hypervisor": "VMware",
>       "hypervisorversion": "4.1",
>       "id": "10"
>     },
>     {
>       "hypervisor": "VMware",
>       "hypervisorversion": "5.0",
>       "id": "11"
>     },
>     {
>       "hypervisor": "VMware",
>       "hypervisorversion": "5.1",
>       "id": "12"
>     },
>     {
>       "hypervisor": "VMware",
>       "hypervisorversion": "5.5",
>       "id": "c57558bf-abc1-11ee-8957-1e00da000152"
>     },
>     {
>       "hypervisor": "VMware",
>       "hypervisorversion": "6.0",
>       "id": "cbaae7e7-abc1-11ee-8957-1e00da000152"
>     },
>     {
>       "hypervisor": "VMware",
>       "hypervisorversion": "6.5",
>       "id": "cd741f46-abc1-11ee-8957-1e00da000152"
>     },
>     {
>       "hypervisor": "VMware",
>       "hypervisorversion": "6.7",
>       "id": "cf92dd65-abc1-11ee-8957-1e00da000152"
>     },
>     {
>       "hypervisor": "VMware",
>       "hypervisorversion": "6.7.1",
>       "id": "cf92e9d9-abc1-11ee-8957-1e00da000152"
>     },
>     {
>       "hypervisor": "VMware",
>       "hypervisorversion": "6.7.2",
>       "id": "cf92f45c-abc1-11ee-8957-1e00da000152"
>     },
>     {
>       "hypervisor": "VMware",
>       "hypervisorversion": "6.7.3",
>       "id": "cf92fed5-abc1-11ee-8957-1e00da000152"
>     },
>     {
>       "hypervisor": "VMware",
>       "hypervisorversion": "7.0",
>       "id": "d14750d0-abc1-11ee-8957-1e00da000152"
>     },
>     {
>       "hypervisor": "VMware",
>       "hypervisorversion": "7.0.1.0",
>       "id": "d1487eae-abc1-11ee-8957-1e00da000152"
>     },
>     {
>       "hypervisor": "VMware",
>       "hypervisorversion": "7.0.2.0",
>       "id": "d2c839eb-abc1-11ee-8957-1e00da000152"
>     },
>     {
>       "hypervisor": "VMware",
>       "hypervisorversion": "7.0.3.0",
>       "id": "d2c96de9-abc1-11ee-8957-1e00da000152"
>     },
>     {
>       "hypervisor": "VMware",
>       "hypervisorversion": "8.0",
>       "id": "d414b048-abc1-11ee-8957-1e00da000152"
>     },
>     {
>       "hypervisor": "VMware",
>       "hypervisorversion": "8.0.0.1",
>       "id": "d414bc3b-abc1-11ee-8957-1e00da000152"
>     },
>     {
>       "hypervisor": "VMware",
>       "hypervisorversion": "8.0.1",
>       "id": "d67c19b1-abc1-11ee-8957-1e00da000152"
>     }
>   ]
> }
> ```
> 
> Import in UI,
> 
>  import-vmw-8002.mp4 
> In Logs,
> 
> ```
> 2024-01-08 09:23:51,845 DEBUG [c.c.a.t.Request] (API-Job-Executor-31:ctx-dfff55c7 job-38 ctx-4a1de5f8) (logid:7f598e47) Seq 1-5056135006653522013: Received:  { Ans: , MgmtId: 167780543, via: 1(10.0.35.115), Ver: v1, Flags: 10, { GetUnmanagedInstancesAnswer } }
> 2024-01-08 09:23:51,845 DEBUG [c.c.a.m.AgentManagerImpl] (API-Job-Executor-31:ctx-dfff55c7 job-38 ctx-4a1de5f8) (logid:7f598e47) Details from executing class com.cloud.agent.api.GetUnmanagedInstancesCommand: 
> 2024-01-08 09:23:51,854 DEBUG [c.c.s.d.GuestOSHypervisorDaoImpl] (API-Job-Executor-31:ctx-dfff55c7 job-38 ctx-4a1de5f8) (logid:7f598e47) Mapping for guest OS ID: rockylinux_64Guest for hypervisor: VMware with version: 8.0.0.2 can not be found. Trying to find one for the parent version: 8.0
> 2024-01-08 09:23:51,862 DEBUG [c.c.u.AccountManagerImpl] (API-Job-Executor-31:ctx-dfff55c7 job-38 ctx-4a1de5f8) (logid:7f598e47) Access granted to Account [{"accountName":"admin","id":2,"uuid":"e06f184a-abc1-11ee-8957-1e00da000152"}] to Service offering {"id":1,"name":"Small Instance","uuid":"5aebc1b7-5e3a-405f-98cf-dec841af8453"}. by AffinityGroupAccessChecker
> ```

thanks @shwstppr 

have you added the hypervisor capabilities for 8.0.0.2 ?
If yes, can  you remove it and retry ?

**Comment by weizhouapache** [comment]
also , it looks the following code snippet is not required if this PR is merged

https://github.com/apache/cloudstack/blob/3f9dd4dc07ff40a4e0216014715e4fe37fc46d28/plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/guru/VmwareVmImplementer.java#L415-L418

**Comment by shwstppr** [comment]
> > Tested `Rocky Linux 9` VM import with VMware 8.0.0.2 hypervisor.
> > ```
> > (localcloud) 🐱 > list hypervisorcapabilities hypervisor=VMware filter=id,hypervisor,hypervisorversion,
> > {
> >   "count": 19,
> >   "hypervisorCapabilities": [
> >     {
> >       "hypervisor": "VMware",
> >       "hypervisorversion": "default",
> >       "id": "8"
> >     },
> >     {
> >       "hypervisor": "VMware",
> >       "hypervisorversion": "4.0",
> >       "id": "9"
> >     },
> >     {
> >       "hypervisor": "VMware",
> >       "hypervisorversion": "4.1",
> >       "id": "10"
> >     },
> >     {
> >       "hypervisor": "VMware",
> >       "hypervisorversion": "5.0",
> >       "id": "11"
> >     },
> >     {
> >       "hypervisor": "VMware",
> >       "hypervisorversion": "5.1",
> >       "id": "12"
> >     },
> >     {
> >       "hypervisor": "VMware",
> >       "hypervisorversion": "5.5",
> >       "id": "c57558bf-abc1-11ee-8957-1e00da000152"
> >     },
> >     {
> >       "hypervisor": "VMware",
> >       "hypervisorversion": "6.0",
> >       "id": "cbaae7e7-abc1-11ee-8957-1e00da000152"
> >     },
> >     {
> >       "hypervisor": "VMware",
> >       "hypervisorversion": "6.5",
> >       "id": "cd741f46-abc1-11ee-8957-1e00da000152"
> >     },
> >     {
> >       "hypervisor": "VMware",
> >       "hypervisorversion": "6.7",
> >       "id": "cf92dd65-abc1-11ee-8957-1e00da000152"
> >     },
> >     {
> >       "hypervisor": "VMware",
> >       "hypervisorversion": "6.7.1",
> >       "id": "cf92e9d9-abc1-11ee-8957-1e00da000152"
> >     },
> >     {
> >       "hypervisor": "VMware",
> >       "hypervisorversion": "6.7.2",
> >       "id": "cf92f45c-abc1-11ee-8957-1e00da000152"
> >     },
> >     {
> >       "hypervisor": "VMware",
> >       "hypervisorversion": "6.7.3",
> >       "id": "cf92fed5-abc1-11ee-8957-1e00da000152"
> >     },
> >     {
> >       "hypervisor": "VMware",
> >       "hypervisorversion": "7.0",
> >       "id": "d14750d0-abc1-11ee-8957-1e00da000152"
> >     },
> >     {
> >       "hypervisor": "VMware",
> >       "hypervisorversion": "7.0.1.0",
> >       "id": "d1487eae-abc1-11ee-8957-1e00da000152"
> >     },
> >     {
> >       "hypervisor": "VMware",
> >       "hypervisorversion": "7.0.2.0",
> >       "id": "d2c839eb-abc1-11ee-8957-1e00da000152"
> >     },
> >     {
> >       "hypervisor": "VMware",
> >       "hypervisorversion": "7.0.3.0",
> >       "id": "d2c96de9-abc1-11ee-8957-1e00da000152"
> >     },
> >     {
> >       "hypervisor": "VMware",
> >       "hypervisorversion": "8.0",
> >       "id": "d414b048-abc1-11ee-8957-1e00da000152"
> >     },
> >     {
> >       "hypervisor": "VMware",
> >       "hypervisorversion": "8.0.0.1",
> >       "id": "d414bc3b-abc1-11ee-8957-1e00da000152"
> >     },
> >     {
> >       "hypervisor": "VMware",
> >       "hypervisorversion": "8.0.1",
> >       "id": "d67c19b1-abc1-11ee-8957-1e00da000152"
> >     }
> >   ]
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Import in UI,
> > import-vmw-8002.mp4
> > In Logs,
> > ```
> > 2024-01-08 09:23:51,845 DEBUG [c.c.a.t.Request] (API-Job-Executor-31:ctx-dfff55c7 job-38 ctx-4a1de5f8) (logid:7f598e47) Seq 1-5056135006653522013: Received:  { Ans: , MgmtId: 167780543, via: 1(10.0.35.115), Ver: v1, Flags: 10, { GetUnmanagedInstancesAnswer } }
> > 2024-01-08 09:23:51,845 DEBUG [c.c.a.m.AgentManagerImpl] (API-Job-Executor-31:ctx-dfff55c7 job-38 ctx-4a1de5f8) (logid:7f598e47) Details from executing class com.cloud.agent.api.GetUnmanagedInstancesCommand: 
> > 2024-01-08 09:23:51,854 DEBUG [c.c.s.d.GuestOSHypervisorDaoImpl] (API-Job-Executor-31:ctx-dfff55c7 job-38 ctx-4a1de5f8) (logid:7f598e47) Mapping for guest OS ID: rockylinux_64Guest for hypervisor: VMware with version: 8.0.0.2 can not be found. Trying to find one for the parent version: 8.0
> > 2024-01-08 09:23:51,862 DEBUG [c.c.u.AccountManagerImpl] (API-Job-Executor-31:ctx-dfff55c7 job-38 ctx-4a1de5f8) (logid:7f598e47) Access granted to Account [{"accountName":"admin","id":2,"uuid":"e06f184a-abc1-11ee-8957-1e00da000152"}] to Service offering {"id":1,"name":"Small Instance","uuid":"5aebc1b7-5e3a-405f-98cf-dec841af8453"}. by AffinityGroupAccessChecker
> > ```
> 
> thanks @shwstppr
> 
> it seems it worked, even without the hypervisor capabilities for 8.0.0.2
> 
> I checked the latest code of main branch, it does have the following code
> 
> https://github.com/apache/cloudstack/blob/3f9dd4dc07ff40a4e0216014715e4fe37fc46d28/engine/schema/src/main/java/com/cloud/hypervisor/dao/HypervisorCapabilitiesDaoImpl.java#L77-L83
> 
> Now I am wondering if #8444 is required.

@weizhouapache we can discuss #8444 separately.
In the original issue, @andrijapanicsb objected to retrieving hypervisor capabilities from the parent version. Also, currently, there is no way to update the minor/patch version capabilities without updating them for the parent version. Therefore, it was discussed to add the contentious versions explicitly for now.
Maybe we can introduce something like https://github.com/apache/cloudstack/compare/main...shapeblue:cloudstack:add-updhycapapi-param?expand=1
I'm happy to close #8444 (or keep just the logs part) if we have a consensus we don't need that

Regarding VMwareVmImplementer change, can we have it in a separate PR? I don't think even with having that we have issues right now

**Comment by weizhouapache** [comment]
> 
> @weizhouapache we can discuss #8444 separately. In the original issue, @andrijapanicsb objected to retrieving hypervisor capabilities from the parent version. Also, currently, there is no way to update the minor/patch version capabilities without updating them for the parent version. Therefore, it was discussed to add the contentious versions explicitly for now. Maybe we can introduce something like https://github.com/apache/cloudstack/compare/main...shapeblue:cloudstack:add-updhycapapi-param?expand=1 I'm happy to close #8444 (or keep just the logs part) if we have a consensus we don't need that
> 
> Regarding VMwareVmImplementer change, can we have it in a separate PR? I don't think even with having that we have issues right now



@shwstppr 
agree
I am good with merging this now.

**Review by andrijapanicsb — APPROVED** [review]
LGTM

**Pull request merged**